### DISCUSSION TITLE
things to do

### DISCUSSION THREAD

**Discussion by mflatt** [racket/rhombus#188] [General]
Some things that we can see currently need to be done, roughly in order of complexity:

* examples – See #186. There's a chicken-and-egg problem here of having enough libraries and figuring out what the libraries should look like.

* fancy-app – Might be a relatively small detail, but better to sort out earlier instead of later.

* unit testing – Something small and simple enough to be in the core, and that composes well with abstractions like functions and modules (instead of having a parallel set of forms for abstracting over tests).

* errors – Figure out how to get primitives to complain with different names, but cheaply. Hopefully, this is not a compiler-improvement exercise.

* pretty printer for shrubberies – Sorawee has the starting point here.

* syntax-parse – Make more syntax-parse functionality available in the pattern and template sublanguages. One particular challenge is figuring out what needs to be in the base language and what can be in a separate library.

* Scribble – Adapted to shrubbery notation (shcribble?), with documentation forms that adapt better to different languages, and with a cleaned-up base library.

* in-source documentation – Possibly along the lines of `scribble/srcdoc`. Depends on Scribble adaptation.

* class system design – Build on the `struct`-like `class` form, but with methods? With `racket/generic` and/or `racket/class` integration?

* core datatypes – Maps, etc. Jack is working on this, but included here in an attempt at completeness and context.

* strings – Support for Unicode grapheme clusters (not just code units). There are questions of how Racket, port counting, and the drawing/GUI/editor libraries interact, so this is an especially wide-ranging problem. But there are some basic design decisions and functionality that might be worked out piecewise, and it might not be all that much code in the end.

* annotations/contracts – Extend the annotation library in the direction of contracts. Similar to syntax-parse, figuring out what needs to be core is part if the challenge here. This one interacts with research and ongoing work by Lukas @ Northwestern.

**Comment by 97jaz** [comment]
@mflatt Regarding Unicode grapheme clusters: Racket uses data files from Unicode 7. The text segmentation spec has changed somewhat since then (mostly to account for emoji, I think, e.g., flags with zero-width joiners). What are your thoughts on moving to Unicode 14? (And do BC and CS need to change in lock step? I notice that they are both using Unicode 7, but I don't know if that was planned.)

**Reply by mflatt** [reply]
Probably it makes sense to move to Unicode 14 as part of a set of unavoidable breaking changes. For example, port counting might switch to grapheme-cluster counting at the same time.

I checked at some point to make sure BC and CS use the same Unicode version. I think it turns out that Chez Scheme and BC were already using the same version, though.

**Reply by 97jaz** [reply]
I have a [PR](https://github.com/racket/racket/pull/4043) up now with two commits for Unicode 14, one for Chez and one for Racket BC. It was surprisingly painless.

What's the next step for strings? The meeting notes mention three points that @samth brought up:
- representation (Unicode scalar values vs. UTF-8)
- (im)mutability
- "deprioritizing indexed access," which is related both to representation and to the preferred view of the data (code units, scalar values, extended grapheme clusters, whatever else)

Compatibility with the existing Racket ecosystem is a big factor here, but I don't know how big. Do Racket strings need to be the same thing as Rhombus strings? If so, then UTF-8 looks like a non-starter.

**Reply by mflatt** [reply]
I thought about possible steps last week, and here are some ideas. I don't know whether any of this would actually work out or be the right idea, though.

 * Keep "string" and "character" as-is.

 * Add `string-grapheme-length`, which takes a string and an offset and reports the length in characters of a grapheme starting from that offset (independent of whatever characters cam before).

   This operation can be used to find graphemes by starting from the beginning of the string. Along those lines, it could also be used to implement `string-grapheme-count`.

 * Is it possible to add ways to combine strings that insert characters (like space) as necessary to prevent joining of graphemes between given strings? For example, appending two strings in this ways would produce a string whose grapheme count is the sum of the counts of the appended strings.

   There are all sorts contexts where you want to preserve grapheme count, and  if it's possible, maybe Rhombus string-formatting things should generally use this view (not a character-based view). For example, the equivalent of `error: ~a` with a string argument should not cause the space before `~a` to be combined with the first part of a string substituted in `~a`.

 * Add `graphemes<?`, `graphemes-ci<?`, etc. (not sure about the names) for comparing string by normalized grapheme instead of characters.

   I don't know whether this is what string equality should use by default in Rhombus, but maybe.

 * Add `read-grapheme` alongside `read-byte` and `read-char`.

   To reduce allocation, return a grapheme as a char when possible (in composed-normalized form), or as a string otherwise.

 * Change port line counting to count positions and columns in graphemes, not characters — either always or by request.

   The "always" option would be backward-incompatible, obviously, but it might make sense to take that leap. The "by request" part would need significant new plumbing in the port layer. We should experiment to see how bad the backward-incompatibility hit would be.

 * Change `draw-text` in non-combining mode to draw graphemes independently, not characters — either always or by request.

   The "by request" part is easier here, so maybe we don't need to break backward compatibility. The Rhombus wrapper should probably turn on grapheme mode, though.

 * Change `snip-snip%` to count by graphemes instead of characters — either always or by request.

   We'll want DrRacket to use the new mode, and that may tip us into backward incompatibility as the right choice overall.

 * Regexps: maybe not worth doing anything.
 
   One possibility would be adding `#ux` alongside `#rx` and `#px`, where `#ux` performs grapheme matching instead of character or byte matching (and don't have `#ux#` byte-string variant). That could work if it's possible to convert an individual grapheme to a character regexp that matches all characters encoding the same grapheme, and that's how it would be implemented. Maybe general regexp matching is just a character thing, though.

**Reply by rfindler** [reply]
These changes related to text% and drracket sound just right to me. I guess the `string-snip%` changes would probably be reasonable to do in a "backwards incompatible" way (as it seems reasonable to declare the current behavior as buggy at the string-snip layer, unlike strings themselves).

I was also wondering about introducing a new kind of value that was intended to play a role like strings but was a more sophisticated data structure that was grapheme based but supported things like constant-time indexing (constant time concatenation anyone? :), but one argument against an approach like that (beyond the library interoperability concerns) is that `string-snip%` wants a more packed representation and the abstraction is already provides might be a better place to implement such things anyway. (At least it seems like the approach of just adding operations instead of a new kind of data would work well for that corner of Racket.)

**Reply by samth** [reply]
I should have made a new discussion before this happened here, but here are a few thoughts.

The big questions for Rhombus, as I see them, are:

1. Are strings mutable? Mutable strings are a bad idea, and it would be nice to not have to deal with them.
2. Is `String` an interface, which could be backed by multiple implementations, or a particular data type which Rhombus defines?
3. Do `String`s support constant-time indexing? The advice given today is usually that arbitrary indexing is never what you want to be doing anyway.
4. What's the default iteration over: grapheme clusters (like Swift), Unicode code points (like Racket), bytes, something else?

And finally:

5. Does `String` in Rhombus mean the same thing as `string?` in Racket?

My preference, in the absence of worrying about compatibility with Racket, is to say that `String` is an fixed immutable data structure represented with UTF-8 and where the default iteration is over EGCs, with the importance in that order.  

This could support `String` being `string?`, but that would require changing `string-ref` in Racket to not always be constant time. (If you record an ASCII-only bit, then those strings can have constant time indexing). That would also require all Rhombus strings to be immutable, so interoperation was less convenient for mutable strings.  Another possible way to make this work would be to have the runtime have two string implementations, with Rhombus creating the UTF-8 versions and the UTF-32 versions supporting constant-time indexing.

**Reply by gus-massa** [reply]
`string-ref` not in constant time like `O(N)` or `O(ln(N))`?

In spite it is not the main objective, there are a lot of benchmarks that assume the the world is in ASCII, and use `string-ref` and `string-set!`.

I think that in real world examples, I only used concatenation and search/regexp.

**Reply by 97jaz** [reply]
Gus wrote:
> `string-ref` not in constant time like O(N) or O(ln(N))?

O(n) in a straightforward implementation.

This is why I wrote that UTF-8 seems like a non-starter if Racket strings and Rhombus strings needs to be the same thing. Changing Racket strings to UTF-8 would turn some linear algorithms in Racket into quadratic ones. But we could improve the time complexity by using a technique similar to what Racket does today for (list) `length` -- by lazily building an index of a string as it's accessed. In fact, Swift's string views do this; they build "[breadcrumbs](https://swift.org/blog/utf8-string/#breadcrumbs)" to amortize what would otherwise be a linear scan.

Sam wrote:
> Another possible way to make this work would be to have the runtime have two string implementations, with Rhombus creating the UTF-8 versions and the UTF-32 versions supporting constant-time indexing.

Yeah, and it would be relatively easy to allow Rhombus to consume either. But although I don't have a good sense of how much code in the runtime system cares about the representation of strings, I'm worried that it's a lot.

**Reply by 97jaz** [reply]
Also:
>  (If you record an ASCII-only bit, then those strings can have constant time indexing). 

You'd think so, but `"\r\n"` is a single grapheme under [Unicode's rules](https://www.unicode.org/reports/tr29/tr29-39.html#GB3).

**Reply by LiberalArtist** [reply]
I haven't had as much time as I'd like to follow the Rhombus discussions, but I definitely have some thoughts about strings: I'm not sure if this is the best place to write them. One fairly brief thought is, when reading things like the [Swift String Manifesto's section on substrings](https://github.com/apple/swift/blob/main/docs/StringManifesto.md#substrings):
> When implementing substring slicing, languages are faced with three options:
>
> 1. Make the substrings the same type as string, and share storage.
> 2. Make the substrings the same type as string, and copy storage when making the substring.
> 3. Make substrings a different type, with a storage copy on conversion to string.

it strikes me that this could be viewed as a garbage collection problem (and Swift, of course, doesn't have a real garbage collector). In the further discussion of the shared-storage option, the Swift developers write:
> This is a simple model with big efficiency gains when chopping up strings into multiple smaller strings. But it does mean that a stored substring keeps the entire original string buffer alive even after it would normally have been released.
>
> This arrangement has proven to be problematic in other programming languages, because applications sometimes extract small strings from large ones and keep those small strings long-term. That is considered a memory leak and was enough of a problem in Java that they changed from substrings sharing storage to making a copy in 1.7.

But does a substring really need to keep the *entire* original string buffer alive? I can imagine a mechanism (whether implemented as a library or with special support from the collector) that, when the original string becomes unreachable, would break up its storage into smaller buffers and allow the unused space to be reclaimed. Perhaps the "breaking up" could happen only when the collector would be moving memory around anyway.

I'm not sure what the state of the art is here for mostly-functional garbage-collected languages. It seems like there could be prohibitive complications. On the other hand, managing the underlying storage in memory seems to apply even if the user-facing string type is a rope, cord, or whatever.

**Reply by gus-massa** [reply]
How bad is storing the strings as in a `hash` full of `graphemes`? (See bad implementation bellow.)
I called them `chain`s that is somewhat similar to `string`s. chain-ref/-set/-set!are O(ln(N)), I added an offset, so `chain-append` is not quadratic and it's not necessary to use `StringBuilder` like in C#. And `subchain` can have more special cases to remove only a few sharacters^H graphemes at the beginning or at the end. (I implemented only two very obvious cases.)

The problems may be that hash are slower than raw access to memory, and I don't know how to connect something like this to the ffi. 


    #lang racket

    (struct chain* (h offset) #:mutable #:transparent)

    (define chain? chain*?)

    (define (string->chain s)
      (define h (make-immutable-hash))
      (for ([c (in-string s)]
            [p (in-naturals)])
        (set! h (hash-set h p c)))
      (chain* h 0))

    (define (chain->string x)
      (define h (chain*-h x))
      (apply string
             (for/list ([n (in-range (chain*-offset x)
                                     (+ (chain*-offset x)
                                        (hash-count h)))])
               (hash-ref h n))))

    (define (chain-length x)
      (hash-count (chain*-h x))) 

    (define (chain-ref x n)
      (hash-ref (chain*-h x) (+ n (chain*-offset x)))) 

    (define (chain-set! x n c)
      (set-chain*-h! x (hash-set (chain*-h x) (- n (chain*-offset x)) c))) 

    (define (chain-set x n c)
      (chain* (hash-set (chain*-h x) (- n (chain*-offset x)) c)) (chain*-offset x)) 

    (define (chain-append* x y)
      (cond
        [(> (chain-length x) (chain-length y))
         (define h (chain*-h x))
         (define o (+ (chain*-offset x) (chain-length x)))
         (for ([n (in-range (chain-length y))])
           (set! h (hash-set h (+ n o) (chain-ref y n))))
         (chain* h (chain*-offset x))] 
        [else
         (define h (chain*-h y))
         (define o (- (chain*-offset y) (chain-length x)))
         (for ([n (in-range (chain-length x))])
           (set! h (hash-set h (+ n o) (chain-ref x n))))
         (chain* h o)]))

    (define (chain-append x . rest)
      (foldl chain-append* x rest))

    (define (subchain x a b)
      (define h (chain*-h x))
      (define o (chain*-offset x))
      (define l (hash-count h))
      (cond
        [(and (= a 1) (= l b))
         (chain* (hash-remove h o) (+ o 1))]
        [(and (= a 0) (= l (+ b 1)))
         (chain* (hash-remove h (- (+ o l) 1)) o)]
        [else
         (define h2 (make-immutable-hash))
         (for ([p (in-range a b)])
           (set! h2 (hash-set h2 (- p a) (hash-ref h (+ p o)))))
         (chain* h2 0)]))
     
    ; Examples
    (define b (string->chain "ba?ana"))
    b
    (chain->string b)
    (chain-ref b 2)
    (chain-set! b 2 #\n)
    (chain->string b)
    (define a (string->chain "apple"))
    (chain->string (chain-append a b))
    (chain->string (chain-append b a))
    (chain->string (chain-append a b a))
    (define ba (chain-append b a))
    (chain->string (chain-append ba b a ba b a))
    (chain->string (chain-append b a b a b a b a))
    (chain->string (chain-append b a ba ba b a))
    (chain->string (subchain ba 0 11))
    (chain->string (subchain ba 1 11))
    (chain->string (subchain ba 0 10))
    (chain->string (subchain ba 1 10))

**Reply by LiberalArtist** [reply]
The discussion about deprioritizing random access (which I definitely endorse—most often I want efficient concatenation and slicing) reminded me of an [old mailing-list thread](https://groups.google.com/g/racket-dev/c/KTNMrc2ahnI/m/ajqXVkvGAwAJ) in which @lexi-lambda shared this benchmark:
```racket
#lang racket/base
(define (catport in out)
  (let ((x (read-char in)))
    (unless (eof-object? x)
      (write-char x out)
      (catport in out)))) 
```
and reported (in 2016):
> It is a very simple cat implementation the benchmarks themselves invoke
it on a file about 4.5MB in size, and it takes about 15 seconds. This is
comparable with some other Scheme implementations, but it’s thoroughly
beaten by Bigloo, Chez, and Larceny, which are all in the 1-3 second
range. Replacing `read-char` and `write-char` with `read-byte` and `write-byte`
brings the time down to about 11 seconds, but it doesn’t change things
significantly.

to which @mflatt replied in part:
> Yes, the port abstraction offers at once too much (in terms of
synchronization primitives and byte counting) and too little (in terms
of a way to define a single-byte or single-char fast path) to make this
microbenchmark go super fast.

Similarly, I remembered comments from @mflatt in [`json` (L200-L203)](https://github.com/racket/racket/blob/93f6fdc7246b00fd02d57a2d4a84ce3b34855313/racket/collects/json/main.rkt#L200-L203) and [`racket/fasl` (L395-L396)](https://github.com/racket/racket/blob/93f6fdc7246b00fd02d57a2d4a84ce3b34855313/racket/collects/racket/fasl.rkt#L395-L396) about avoiding ports for highly tuned performance. I think we should consider how strings can interact with other types like ports to support solving such problems with code that is both elegant and efficient.

**Reply by 97jaz** [reply]
@LiberalArtist I've been thinking about ports recently, too, originally in the context of grapheme-level I/O, but that led me to considerations similar to yours. Maybe it would be a good time to tease apart the different concerns that are now fused together in the port abstraction? This probably warrants its own discussion. (I should have started one specific to strings, too.)

**Reply by rocketnia** [reply]
In hand-rolled parsers I've written JavaScript, I've used regexes quite a bit so that I can allow regex internals to do the iteration work without having to hand off to JavaScript function calls. And that's a habit I've carried over into Racket.

In a language like Rust or C where loops are cheap enough already, using regexes for performance may be backwards. I'm not sure exactly where Racket or Rhombus will fall on that spectrum in the long term.

Anyway, the architecture of strings and ports for performance seems to be deeply related to regexes, as long as regexes serve a performance purpose like this. And for this purpose, regexes might be a bit underpowered; they can only result in a matched snippet, and they can't really resume what they were doing after that snippet is consumed. For performant transduction between input ports and output ports, and between ports and streams, there might be a need for a more extended system than mere regexes.

A customizable DSL for such a purpose would probably favor a lot of static typing and inlining (to approach Rust or C performance), but even a DSL made up of a zoo of built-in combinators would probably go a long way.

On a bit of a tangent, a few weeks ago I was reading up on [suffix trees](https://en.wikipedia.org/wiki/Suffix_tree) and [suffix arrays](https://en.wikipedia.org/wiki/Suffix_array) on Wikipedia, and I found this to be a potentially interesting sweet spot:

> One of the first algorithms to achieve all goals is the SA-IS algorithm of Nong, Zhang & Chan (2009). The algorithm is also rather simple (< 100 LOC) and can be enhanced to simultaneously construct the [LCP array](https://en.wikipedia.org/wiki/LCP_array). The SA-IS algorithm is one of the fastest known suffix array construction algorithms.

So besides ports and strings, in the long run there might be a use for an "indexed string" data type where the suffix array and LCP array have already been constructed. This could then be the recommended string type for conducting global regex searches and such.

I don't suppose suffix arrays would be useful for improving the performance of `cat` or `#lang` tokenizers, but they're a bit of a window into what other possible concerns there are in the design of string internals.

**Reply by gus-massa** [reply]
> deprioritizing random access

I agree with that, but for example most people will expect that peeking at the last character is cheap (like O(1) or O(ln(N)))

For example check that the last character is a space to implement a custom version of `trim`, or a colon to check if the user want to write a letter. Most cases I have in mind, it's a short text, like 20 characters, but it may be something longer like a big s-expression or a html document (assuming nobody writes after the `</html>` tag).

**Reply by jackfirth** [reply]
> What's the default iteration over: grapheme clusters (like Swift), Unicode code points (like Racket), bytes, something else?

On this question, might I humbly suggest not having a default iteration order at all? Instead we could have `string.chars()`, `string.bytes(encoding)`, and `string.graphemeClusters()` return iterable views of the string. They could even be collection views.

**Reply by mflatt** [reply]
As an initial experiment, I've tried adding core grapheme-cluster operations with a focus on efficiently streaming characters (i.e., code points) through a grapheme tracker:

https://github.com/mflatt/racket/commit/e1fe2b28d8a841e95313134d66fd9ada9549f762

I also tried updating CS to count port positions in terms of Unicode grapheme clusters instead of Unicode code points:

https://github.com/mflatt/racket/commit/3014238259ab3614e156cb0ae543d005db791e27

That second commit's message has some numbers on the cost of always counting in terms of graphemes instead of code points. It's significant, but probably reasonable. I have to use large inputs to provoke deltas on the order of 10 ms.

(None of this is for Racket v8.6, of course. If it goes well, it could be in v8.7.)

**Reply by samth** [reply]
Some of the BC numbers there are very surprising. Eg, why is `read-bytes` and `read-string` so much faster than on Racket CS in the old version, when `read-char` is so much slower?

**Reply by samth** [reply]
My other initial thought is that moving to a string representation that was more efficient for ASCII-only strings (such as just using UTF-8 always, my preference, or a new string representation that was used for ASCII-only) at the same time as this change would likely make the overall result a performance improvement, even if the grapheme-cluster changes by themselves introduces overhead.

**Reply by mflatt** [reply]
The `read-char` and `read-byte` versions are Racket loops that have to call into `read-char` and `read-byte`. Calling into non-inlined primitives is relatively slow on BC.

Meanwhile, `read-bytes` and `read-string` have all the counting work built in, which means staying in C on BC. The numbers where BC is 10x as fast are just wrong; I transcribed the wrong columns in a larger set of output, and those should be 2x. The ones where it's 2x or 3x as fast are because the C code is currently faster than the Scheme code. Partly, that speed is obtained in BC by just giving up when positions move out of fixnum range, while CS will keep counting. Partly, even though Chez Scheme generates fast code, I'm still able to even faster code via GCC/Clang for small-integer manipulation.

**Reply by samth** [reply]
Ok, that all makes sense. Can we get back some of the speed by assuming fixnums, at least on 64-bit platforms?

**Reply by mflatt** [reply]
> Can we get back some of the speed by assuming fixnums, at least on 64-bit platforms?

It looks like I was wrong about fixnum aithmetic being relevant. Although I can make the generated code substantially simpler by using fixnum arithmetic, it doesn't run any faster (maybe because the CPU's branch predictor works).

**Reply by samth** [reply]
It might be worth testing that on x64 too, if you haven't.

**Reply by mflatt** [reply]
Some notes after further implementation at racket/racket#4350, racket/draw#49, and racket/gui#277:

> Is it possible to add ways to combine strings that insert characters (like space) as necessary to prevent joining of graphemes between given strings?

No, this idea doesn't work, at least not for the purpose of appending a string of grapheme length N with another of grapheme length M and getting a string that is always N+M graphemes.

> Add `graphemes<?`, `graphemes-ci<?`, etc. [...]
> Add `read-grapheme` [...]

I haven't needed these, so far, and I haven't tried to add them.

> Change port line counting to count positions and columns in graphemes, not characters — either always or by request.

Current implementation is "always".

> Change draw-text in non-combining mode to draw graphemes independently, not characters — either always or by request.

Current implementation is "by request".

> Change `string-snip%` to count by graphemes instead of characters — either always or by request.

Changing "item" in the editor's terminology to mean graphemes instead of characters does not seem to be workable. The editor API and uses are too intertwined with Racket strings and characters. (After arriving at that conclusion, I considered again the possibility of changing Racket's notion of "character" to grapheme, but only briefly, because it would be a huge amount of effort for a relatively small payoff.)

Instead, the current implementation adds grapheme counting to the editor classes and offers methods for converting between grapheme and character positions. Old methods continue to work in terms of characters, and there is no guarantee of grapheme integrity; for example, you can programmatically position the caret inside a grapheme and insert characters, even though the default implementation of arrow-key navigation would not let a user do that. Instead of having the editor "just work" when it comes to graphemes, we'll need to apply incremental changes to various clients of the editor API to make them do the right thing.

> Regexps: maybe not worth doing anything.

"Not worth it" still sounds right.

**Comment by 97jaz** [comment]
> Is it possible to add ways to combine strings that insert characters (like space) as necessary to prevent joining of graphemes between given strings? For example, appending two strings in this ways would produce a string whose grapheme count is the sum of the counts of the appended strings.
>
> There are all sorts contexts where you want to preserve grapheme count, and if it's possible, maybe Rhombus string-formatting things should generally use this view (not a character-based view). For example, the equivalent of error: ~a with a string argument should not cause the space before ~a to be combined with the first part of a string substituted in ~a.

That's a good question. From what I can tell there isn't a code point that extends a cluster on one side and prevents extension on the other. So, for example: `"e\u200C\u0301"` has only 1 grapheme, because the zero-width non-joiner in the middle extends the cluster started by "e," and then the combining acute accent attaches as it normally would. On the other hand, it you replace the zero-width non-joiner with a zero-width space (`\u200B`), you get three graphemes, because the space doesn't combine with either the left or the right side.

I found a [post](https://forums.swift.org/t/strings-in-swift-4/4939) where a member of the Swift core team mentions this issue. Here's the relevant bit:

> It was always well understood that if strings were treated as sequences of
> `UnicodeScalar`s, algorithms such as `lexicographicalCompare`, `elementsEqual`,
> and `reversed` would produce nonsense results. Thus, in Swift 1.0, `String` was
> a collection of `Character` (extended grapheme clusters). During 2.0
> development, though, we realized that correct string concatenation could
> occasionally merge distinct grapheme clusters at the start and end of combined
> strings.
> 
> This quirk aside, every aspect of strings-as-collections-of-graphemes appears to
> comport perfectly with Unicode. We think the concatenation problem is tolerable,
> because the cases where it occurs all represent partially-formed constructs. The
> largest class—isolated combining characters such as ◌́ (U+0301 COMBINING ACUTE
> ACCENT)—are explicitly called out in the Unicode standard as
> “[degenerate](http://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries)” or
> “[defective](http://www.unicode.org/versions/Unicode9.0.0/ch03.pdf)”. The other
> cases—such as a string ending in a zero-width joiner or half of a regional
> indicator—appear to be equally transient and unlikely outside of a text editor.
> 
> Admitting these cases encourages exploration of grapheme composition and is
> consistent with what appears to be an overall Unicode philosophy that “no
> special provisions are made to get marginally better behavior for… cases that
> never occur in practice.”[2] Furthermore, it seems
> unlikely to disturb the semantics of any plausible algorithms. We can handle
> these cases by documenting them, explicitly stating that the elements of a
> `String` are an emergent property based on Unicode rules.

So, kind of a `¯\_(ツ)_/¯` response, really. (I think it's a good idea to consider that these "cases that never occur in practice" might well occur if they're intentionally sought by a malicious user.)

**Reply by rfindler** [reply]
+1 to "(I think it's a good idea to consider that these "cases that never occur in practice" might well occur if they're intentionally sought by a malicious user.)"

**Comment by spdegabrielle** [comment]
> fancy-app – Might be a relatively small detail, but better to sort out earlier instead of later.

What is **fancy-app** ?

**Reply by Metaxal** [reply]
> > fancy-app – Might be a relatively small detail, but better to sort out earlier instead of later.
> 
> What is **fancy-app** ?

https://pkgs.racket-lang.org/package/fancy-app
